include/mqtt/shared_any.hpp

```diff
@@ -0,0 +1,30 @@
+// Copyright Takatoshi Kondo 2019
+//
+// Distributed under the Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(MQTT_SHARED_ANY_HPP)
+#define MQTT_SHARED_ANY_HPP
+
+#include <memory>
+
+#include <boost/type_erasure/any.hpp>
+
+namespace mqtt {
+
+namespace mpl = boost::mpl;
+
+template <typename Concept>
+class shared_any : public boost::type_erasure::any<Concept, boost::type_erasure::_self&> {
```

Comment:
I'm not sure I understand the goal here.

Can't std::any / boost::any hold a shared_ptr ?

Reply:
This is very difficult to explain. The goal is erasing type.
`boost::type_erasure::any` can call member functions without cast. See `type_erased_socket.hpp`.

It is one of runtime type erasure.

Reply:
https://www.boost.org/doc/libs/1_70_0/doc/html/boost_typeerasure.html is document.

Boost type erasure doesn't support `shared_ptr` directly. I need to adapt the pointee object that is held by `shared_ptr` to the concept.

The 2nd argument of `boost::type_erasure_any` is to support reference semantics.